cmd/storage-rest-server.go

```diff
@@ -0,0 +1,358 @@
+/*
+ * Minio Cloud Storage, (C) 2016 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"path"
+	"strconv"
+
+	"net/http"
+
+	"encoding/gob"
+	"encoding/hex"
+
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/minio/minio/cmd/logger"
+)
+
+// To abstract a disk over network.
+type storageRESTServer struct {
+	storage *posix
+}
+
+func (s *storageRESTServer) writeErrorResponse(w http.ResponseWriter, err error) {
+	w.WriteHeader(http.StatusForbidden)
+	w.Write([]byte(err.Error()))
+}
+
+// IsValid - To authenticate and verify the time difference.
+func (s *storageRESTServer) IsValid(w http.ResponseWriter, r *http.Request) bool {
+	requestTimeStr := r.Header.Get("X-Minio-Time")
+	requestTime, err := time.Parse(time.RFC3339, requestTimeStr)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return false
+	}
+	utcNow := time.Now()
+	delta := requestTime.Sub(utcNow)
+	if delta < 0 {
+		delta = delta * -1
+	}
+	if delta > DefaultSkewTime {
+		s.writeErrorResponse(w, fmt.Errorf("client time %v is too apart with server time %v", requestTime, utcNow))
+		return false
+	}
+	return true
+}
+
+// DiskInfoHandler - returns disk info.
+func (s *storageRESTServer) DiskInfoHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	info, err := s.storage.DiskInfo()
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = gob.NewEncoder(w).Encode(info)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// MakeVolHandler - make a volume.
+func (s *storageRESTServer) MakeVolHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	err := s.storage.MakeVol(volume)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// ListVolsHandler - list volumes.
+func (s *storageRESTServer) ListVolsHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	infos, err := s.storage.ListVols()
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = gob.NewEncoder(w).Encode(&infos)
+	if err != nil {
+		s.writeErrorResponse(w, err)
```

Comment:
once writing to w we cannot write error again, it will throw multiple write calls.